sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -28,43 +28,12 @@ object Canonicalize {
       e: Expression,
       f: PartialFunction[Expression, Seq[Expression]]): Seq[Expression] = e match {
     case c if f.isDefinedAt(c) => f(c).flatMap(gatherCommutative(_, f))
-    case other => reorderCommutativeOperators(other) :: Nil
+    case other => other.canonicalized :: Nil
```

Comment:
I think this is the key change. This means that we will first reorder leaf-most adjacent commutative operators, and then do this recursively bottom-up.

Reply:
It's probably better to have a new trait `CommutativeExpression` and remove this object.

Reply:
Agreed, fixed in https://github.com/apache/spark/pull/37851/commits/ff2fe3273204da330a2f9aee4d32835749c31b77